controllers/propagator/propagation.go

```diff
@@ -555,8 +560,13 @@ func (r *Propagator) handleDecisions(
 
 			processedResults++
 
-			for refObject := range result.TemplateRefObjs {
-				allTemplateRefObjs[refObject] = true
+			if checkFullSpec {
+				// NOTE: new replicated policies may still appear when we aren't checking the full specs,
+				// which *could* bring new template references... will that be resolved by an individual
+				// replicated policy reconciler with a watch source for those?
```

Comment:
Yeah, this is a problem. The dependency watch library needs the ability to append watches.

I think you are on the right track with the root policy handler. This would just be for definition updates of the root policy. The issue is that an update to a replicated policy, even in this new controller, actually triggers a reconcile on the root policy handler from what I can tell from the code. When we added the root policy status controller, this was a point of debate of if we should ignore changes to the replicated policies and we decided not to in the event a user accidentally or maliciously modified a replicated policy. This is a detail I had forgotten about that a reconcile on the root policy that updates a replicated policy actually triggers two total reconciles. That's another inefficiency to address.

Instead of splitting out the placement updates to be a reconciler on the root policy, I think it should be a reconcile on a replicated policy/cluster decision. The following would cause reconciles:

- The placement/placement rule/placement decision binding is created, updated, or deleted. A reconcile request is made for every new/deleted cluster decisions. This could be based on existing propagated policies in the cache to see what was there previously. Alternatively, you could cache that the cluster decisions per root policy in memory.
- The root policy reconciler still handles root policy definition changes, but it doesn't actually perform the work of building replicated policies. We would add concurrency to the new propagated policy reconciler and remove the custom concurrency of handling decisions in the root policy reconciler. The root policy reconciler would have an event source (i.e. Go channel) that feeds reconcile requests to the new propagated policy reconciler. This root policy reconciler would still add filler status for new replicated policies to be created to not cause the root policy status controller to update the root policy status for every new replicated policy being created by the new replicated policy reconciler. Perhaps there is a more clever way here.
- A change to a replicated policy but we need to find a way to distinguish between a user change and a controller change. Perhaps the managed fields could allow us not to duplicate the request if the last update was from this controller.

The following additional changes would be needed:
- All custom locking is removed since deduplication of requests is already handled by controller-runtime's queue.
- The hub template watches would be per replicated policy and trigger a reconcile on the replicated policy. This would solve one of our other inefficiencies.
- You could optionally optimize for the case where selective policy enforcement is used to not reexecute templates but I think that is not needed.

Reply:
I think you're right: a per-replicated-policy reconcile will be better. I was hoping that the double-reconcile issue here (where this new reconciler does a change which just triggers the root-policy reconcile) could be addressed by a replicated-policy reconciler anyway, but now I see that would still be unnecessary duplication of work. Even with additional changes/fixes, it doesn't seem like my approach will gain us much.

I'm worried that calculating what replicated policies need reconciles based on placement changes will put more work in the `EventHandler` / mapper functions than is intended by controller-runtime, but it's hard for me to tell if that's actually a valid concern.

I think a new bottleneck could come up which is that each replicated policy reconcile will need to calculate its correct decision and any remediationAction overrides... From sketching out different cases of changes, I'll only feel confident it can get it right if it looks through all of the relevant bindings, placements, and decisions for its root policy. But, those will all be in the cache, so it's not a *terrible* problem. It's just that the current approach nicely calculates that for all the replicated policies at once.

The double-reconcile problem will still appear. A placement change will trigger a reconcile on the replicated policy, and if that makes any updates it will trigger a new reconcile. I don't see a good way to filter that out, maybe it's a necessary evil, or we can solve it later. It's still better than an update to one replicated policy effectively causing a re-reconcile for all of the replicated policies.

I don't think a replicated-policy-reconciler would be able to skip some template resolution. But I think with the identified improvements to `go-template-utils`, that will be ok: [ACM-7398](https://issues.redhat.com/browse/ACM-7398), [ACM-7402](https://issues.redhat.com/browse/ACM-7402).

Reply:
So the new proposed reconciler would watch:
- placement stuff (like the one in this PR, but map to replicated policies)
- replicated policies
- dependency-watches for objects used in a replicated policy's template resolution
- a source that would come from the root-policy-controller telling it that the root spec has changed.

I think we need to figure out where the PolicySet changes reconcile to... Originally I had that in this proposed reconciler, but now I forget why I removed that.

Reply:
@JustinKuli, I just checked and we can't use managed fields like I originally thought to detect the double reconcile.

Perhaps you could keep a cache of resourceVersion entries for the last handled replicated policies? Then a predicate on watching replicated policies could use it.

This would work because dependency watches and the root policy controller source would bypass this check.